firmware/src/jtag.rs

```diff
@@ -244,18 +243,19 @@ impl<'a> JTAG<'a> {
                     return;
                 }
 
-                // Set TDI, read TDO, and toggle TCK.
+                // We set TDI half a period before the clock rising edge where it is sampled
+                // by the target, and we sample TDO immediately before the clock falling edge
+                // where it is updated by the target.
```

Comment:
I'm not sure this TDO sampling approach is correct, but I also don't see when it could cause problems.

Reply:
The JTAG spec just says that "Changes in the state of the signal driven through TDO shall occur only on the falling edge of either TCK or the optional TRST", without any requirement on when it's sampled. This way we give ourselves the maximum possible time for TDO to settle. In a hardware implementation you'd want to sample on the rising edge in synchronous logic, but in software we can just as easily wait longer.

Still I agree it looks a bit unusual compared to sampling on the rising edge! I can't think of any reasons this wouldn't be fine, but we could sample on the rising edge just to be conservative.

Reply:
Ah, I see. Thanks for clarifying!